indico/modules/events/editing/client/js/components/FileManager/reducer.js

```diff
@@ -0,0 +1,109 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2019 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import _ from 'lodash';
+import {combineReducers} from 'redux';
+import * as actions from './actions';
+
+export default combineReducers({
+  fileTypes: (state = [], action) => {
+    const fileType = _.find(state, {id: action.fileTypeId});
+    switch (action.type) {
+      case actions.MARK_UPLOADED:
+        action.file.state = 'added';
+        if (fileType.allowMultipleFiles) {
+          fileType.files = fileType.files.concat(action.file);
+        } else {
+          fileType.files = [action.file];
+        }
+        return [...state];
```

Comment:
do we need to clone the state here?

Reply:
It's just to signal that the state has changed, as a shallow comparison will not work.

Reply:
that implies mutating state ;)

Reply:
Not sure I understand.

Reply:
Is mutating state more acceptable in `useReducer` than with redux?

Because if you don't mutate the old state then by definition the new state you return is either the same (if nothing changed) or a new object (if something changed).

Reply:
Sure. But then instead of "cloning" the state we should build a new state object. Anyway, I agree that we should not mutate the state. We just did it like that initially because it becomes annoying once you have nested states.